NS1 doesn't support region OC. Handle it explicitly in the provider

pavantc opened the pull request:
NS1 does not handle region 'OC' (Oceania) which is a valid region that could be part of the rules in dynamic DNS configuration.

The strategy used in this fix is to handle 'OC' specified in the rules and convert it to a list of countries using the corresponding country codes (which is definitely supported in NS1).  This is handled in both directions - while pushing data to NS1 and while reading/interpreting response from NS1:

1. When pushing to NS1:
Any occurrence of 'OC' in the dynamic rules are converted to a list of countries before pushing to NS1

2. In the response:
Presence of *all* the countries that are part of 'OC' is detected and translated back to 'OC' region in the provider before presenting the response to the application. Accordingly, the entries under the 'country' tag are removed (since we are replacing those with 'OC').

The other part of the PR is related to testing the added functionality.

pavantc commented on tests/test_octodns_provider_ns1.py:
```diff
@@ -1092,6 +1135,20 @@ def test_data_for_dynamic_A(self):
+        oc_country_list = self._CONTINENT_TO_LIST_OF_COUNTRIES['OC']
+        ns1_record['regions']['lhr']['meta']['country'] = oc_country_list
+        data3 = provider._data_for_A('A', ns1_record)
+        assert('OC' in data3['dynamic']['rules'][0]['geos'])
```
I probably should add another assert here that none of the 'OC' countries are present under the 'country' tag.

ross replied:
In tests you'll want to use `self.assert*` rather than bare `assert` so that you get all the unittest features/details.

```suggestion
        self.assertTrue('OC' in data3['dynamic']['rules'][0]['geos'])
```

ross commented on octodns/provider/ns1.py:
```diff
@@ -351,12 +358,35 @@ def _data_for_dynamic_A(self, _type, record):
 
+            if len(oc_countries):
+                all_oc_countries = self._CONTINENT_TO_LIST_OF_COUNTRIES['OC']
+                if sorted(oc_countries) == sorted(all_oc_countries):
```
If you change `oc_countries` and `_CONTINENT_TO_LIST_OF_COUNTRIES` to be `set`s rather than `list`s you can just compare them directly here. Should be simpler/cleaner than having to sort.

ross commented on octodns/provider/ns1.py:
```diff
@@ -351,12 +358,35 @@ def _data_for_dynamic_A(self, _type, record):
+
+                if con == 'OC':
+                    oc_countries.append(country)
                 geos.add('{}-{}'.format(con, country))
```
If you do an `else` here and only add non-OC things to `geos` you can avoid having to iterate and remove them down below. The section below can just have an `else` for "not all OC" and in that else loop over what you do have and add them in. It's minor nits, but feels a bit cleaner than adding them and then maybe removing them later.

ross commented on tests/test_octodns_provider_ns1.py:
```diff
@@ -521,6 +521,12 @@ def test_data_for_CNAME(self):
+        'OC': ['FJ', 'NC', 'PG', 'SB', 'VU', 'AU', 'NF', 'NZ', 'FM', 'GU',
+               'KI', 'MH', 'MP', 'NR', 'PW', 'AS', 'CK', 'NU', 'PF', 'PN',
+               'TK', 'TO', 'TV', 'WF', 'WS'],
+    }
```
Better to just refer to this as `Ns1Provider. _CONTINENT_TO_LIST_OF_COUNTRIES` rather than duplicate it here.

ross commented on tests/test_octodns_provider_ns1.py:
```diff
@@ -926,6 +932,43 @@ def test_monitors_gc(self, monitors_for_mock, datafeed_delete_mock,
+        rule0['geos'] = ['OC']
+        ret, _ = provider._params_for_A(self.record)
+        self.assertEquals(sorted(ret['regions']['lhr']['meta']['country']),
+                          sorted(self._CONTINENT_TO_LIST_OF_COUNTRIES['OC']))
```
If the provider stuff is updated to sets you should be able to do something like the following here.

```suggestion
        self.assertEquals(set(ret['regions']['lhr']['meta']['country']),
                          self._CONTINENT_TO_LIST_OF_COUNTRIES['OC'])
```

ross commented on tests/test_octodns_provider_ns1.py:
```diff
@@ -1092,6 +1135,20 @@ def test_data_for_dynamic_A(self):
+        ns1_record['regions']['lhr']['meta']['country'] = partial_oc_cntry_list
+        data4 = provider._data_for_A('A', ns1_record)
+        for c in partial_oc_cntry_list:
+            assert('OC-{}'.format(c) in data4['dynamic']['rules'][0]['geos'])
```
```suggestion
            self.assertTrue('OC-{}'.format(c) in data4['dynamic']['rules'][0]['geos'])
```

pavantc commented:
Addressed all comments @ross. Agree to all of them. (Didn't accept suggestions on the web interface since I needed to make some changes anyway, so limited it to one commit).
